src/Lykke.AlgoStore.MatchingEngineAdapter.Services/Listening/ConsumerLoadBalancer.cs

```diff
@@ -0,0 +1,110 @@
+        private readonly IMessageQueue _requestQueue;
+        private readonly IMatchingEngineAdapter _matchingEngineAdapter;
+        private readonly ILog _log;
+
+        private Thread _balancingThread;
+        private int _loadCounter;
+        private bool _isDisposed;
+
+        /// <summary>
+        /// Initializes a <see cref="ConsumerLoadBalancer"/> with a given <see cref="IMessageQueue"/> to process messages from
+        /// </summary>
+        /// <param name="requestQueue">A <see cref="IMessageQueue"/></param>
+        /// <param name="matchingEngineAdapter">A <see cref="IMatchingEngineAdapter"/></param>
+        /// <param name="log">A <see cref="ILog"/></param>
+        /// <exception cref="ArgumentNullException">Thrown when <paramref name="requestQueue"/> is null</exception>
+        public ConsumerLoadBalancer(IMessageQueue requestQueue, IMatchingEngineAdapter matchingEngineAdapter, [NotNull] ILog log)
+        {
+            _requestQueue = requestQueue ?? throw new ArgumentNullException(nameof(requestQueue));
+            _matchingEngineAdapter = matchingEngineAdapter ?? throw new ArgumentNullException(nameof(matchingEngineAdapter));
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+
+            _workers.Add(new ConsumingWorker(_requestQueue, _matchingEngineAdapter, _log));
+            _balancingThread = new Thread(DoBalancing);
+            _balancingThread.Start(_cts.Token);
+        }
+
+        /// <summary>
+        /// Disposes this <see cref="ConsumerLoadBalancer"/> and all of the <see cref="ConsumingWorker"/> it manages
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        /// <summary>
+        /// Balancing thread which handles creating/removing instances of <see cref="ConsumingWorker"/>
+        /// based on the average load in the past 10 seconds
+        /// </summary>
+        /// <param name="cancellationTokenObj">A <see cref="CancellationToken"/> used for stopping the thread</param>
+        private void DoBalancing(object cancellationTokenObj)
+        {
+            var cancellationToken = (CancellationToken)cancellationTokenObj;
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var avg = _workers.Average(c => c.LastMessageTime);
+
+                var now = DateTime.Now.Ticks;
+
+                // 1 tick: 100 ns; 500,000 ticks: 50 ms
+                if (now - avg < 500_000)
+                    _loadCounter++; // Use the load counter to check how many consecutive times the load has been high (or low)
+                else if (now - avg > 10_000_000) // 10M ticks: 1 second
+                    _loadCounter--; // Decrement if the load is low
+
+                if (_loadCounter >= 4) // If the load is high for 2 seconds straight, spin up a new consumer
+                {
+                    _workers.Add(new ConsumingWorker(_requestQueue, _matchingEngineAdapter, _log));
```

Comment:
Each ConsumingWorker has its own thread. How many threads can this class allocate in total and how efficient it will?